File: 739.每日温度.cpp
/*
 * @lc app=leetcode.cn id=739 lang=cpp
 *
 * [739] 每日温度
 */
#include <iostream>
#include <stack>
#include <vector>
using namespace std;
// @lc code=start
class Solution {
public:
  /**
   * @brief
   * 47/47 cases passed (204 ms)
   * Your runtime beats 5.14 % of cpp submissions
   * Your memory usage beats 67.54 % of cpp submissions (86.7 MB)
   * @param temperatures
   * @return vector<int>
   */
  vector<int> dailyTemperatures(vector<int> &temperatures) {
    auto answer = vector<int>(temperatures.size(), 0);
    auto stack = std::stack<int>();
    stack.push(0);
    for (int i = 1; i < temperatures.size(); i++) {
      for (; !stack.empty() && temperatures[i] > temperatures[stack.top()];) {
        answer[stack.top()] = i - stack.top();
        stack.pop();
      }
      stack.push(i);
    }
    return answer;
  }
};
// @lc code=end

File: 739.每日温度.java
/*
 * @lc app=leetcode.cn id=739 lang=java
 *
 * [739] 每日温度
 */
import java.util.Stack;

// @lc code=start
class Solution {

  /**
   * 47/47 cases passed (120 ms)
   * Your runtime beats 54.87 % of java submissions
   * Your memory usage beats 60.76 % of java submissions (53.1 MB)
   * @param temperatures
   * @return
   */
  public int[] dailyTemperatures(int[] temperatures) {
    var ans = new int[temperatures.length];
    var stack = new Stack<Integer>();
    stack.push(0);
    for (int i = 0; i < temperatures.length; i++) {
      for (
        ;
        !stack.empty() && temperatures[i] > temperatures[stack.lastElement()];
      ) {
        ans[stack.lastElement()] = i - stack.pop();
      }
      stack.push(i);
    }
    return ans;
  }
}
// @lc code=end
